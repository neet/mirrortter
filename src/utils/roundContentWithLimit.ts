import { getTweetLength } from 'twitter-text';
import { config } from '../conifg';

/**
 * Transform Mastodon status to Twitter compatible form
 * @param content Content of a status
 * @param additionalContents Additional text append
 */
export const roundContentWithLimit = (content: string, additionalContents: string[], limit = 240): string => {
  const { ellipsis } = config;
  const joinedAdditionalContents = additionalContents.join(' ');

  let roundedContent = '';

  const contentLength = getTweetLength([
    content,
    joinedAdditionalContents,
    ellipsis,
  ].join(' '));

  const shortageLength = limit - getTweetLength([joinedAdditionalContents, ellipsis].join(' '));

  if (contentLength > limit) {
    roundedContent = [
      content.substr(0, shortageLength),
      joinedAdditionalContents,
      ellipsis,
    ].join(' ');
  } else {
    roundedContent = [
      content,
      joinedAdditionalContents,
    ].join(' ');
  }

  return roundedContent;
};
