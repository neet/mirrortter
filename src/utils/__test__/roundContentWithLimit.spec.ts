import { roundContentWithLimit } from '../roundContentWithLimit';

describe('roundContentWithLimit', () => {

  it ('Rounds text when the content is bigger than the limit', () => {
    const result = roundContentWithLimit('1234567890'.repeat(30), ['additional'], 240);

    expect(result).toMatchSnapshot();
  });

  it ('Won\'t round text in the content is less than or equal to the limit', () => {
    const result = roundContentWithLimit('1234567890', ['additional'], 240);

    expect(result).toMatchSnapshot();
  });
});
