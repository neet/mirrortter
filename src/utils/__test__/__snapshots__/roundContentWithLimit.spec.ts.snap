// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`roundContentWithLimit Rounds text when the content is bigger than the limit 1`] = `"123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890 additional ..."`;

exports[`roundContentWithLimit Won't round text in the content is less than or equal to the limit 1`] = `"1234567890 additional"`;
